core/src/main/java/org/apache/accumulo/core/fate/ZooStore.java

```diff
@@ -164,8 +164,10 @@ public long reserve() {
             }
 
             if (deferred.containsKey(tid)) {
-              if (deferred.get(tid).elapsed().compareTo(Duration.ZERO) > 0) {
-                deferred.remove(tid);
+              long deferredTime = deferred.get(tid);
+              long currentTime = System.nanoTime();
+              if (currentTime >= deferredTime) {
```

Comment:
Its not correct to directly compare two nanoTimes.  This is one of the reasons the nano time class was created, to ensure the comparisons were always done correctly.   The other purpose of the class was to ensure unit conversions were correct ensuring nanos were not compared to millis for example.  Why not use the new class here?

Think this should be the following to be correct.

```suggestion
              if ((currentTime - deferredTime) >= 0) {
```

Reply:
I was hoping to use the new `Timer` class here but ZooStore was relying on the, now removed, `nanoTime.subtract()` and `NanoTime.nowPlus()` methods. I could not think of a clean way to perform this functionality with the Timer class so I just changed it to use long.

ZooStore was the only place where these two methods of NanoTime were being used so I thought it would be okay to remove those. They could either be added back or if someone else can think of a good way to use `Timer` here that would be even better.

Reply:
What do think of adding the following to Timer?

```java

 private Timer(long offset) {
    this.startNanos = System.nanoTime() + offset;
  }

/**
 *  Creates a timer with an offset from the current time.
  */
public static startNew(Duration offset){
    return new Timer(offset.toNanos());
}
```

If we had that then could track timers and create them with the offset.  This would cause the elapsed time to be negative until the offset is reached.

```java
deferred.put(tid, Time.startNow(deferTime));
```

Once the elapsed time for a timer is no longer negative then we know its expired

```java
Timer deferredTime = deferred.get(tid);
if(deferredTime.hasElapsed(0, TimeUntil.MILLIS) {
   // this timer has reached the offset
}
```

For the case of finding the minimum sleep time could do the following.

```
           // A timer with a negative elapsed time has not reached the offset and still has time to sleep.  A timer with a positive elapsed time has surpassed its offset and does not need to sleep.
            long maxElapsed = deferred.values().stream()
                  .mapToLong(timer -> timer.elapsed(MILLIS)).max().orElseThrow();
            if(maxElapsed < 0) {
                this.wait(maxElapsed * -1);
            }
```

Reply:
That seems like a good idea to me. I'll open a quick PR for 2.1 to add this and we can use it here once its merged up.

Reply:
To me, this seems very confusing. It leads to nonsensical interpretations like having a negative elapsed time. To me, this indicates the calling code could be improved with better semantics. If we want a Countdown timer instead of an Elapsed timer, then we should just create a separate timer that counts down, with explicit semantics in its its API, rather than try to recreate it conceptually via low-level manipulations of the internals to get the desired outcome.

Reply:
>  If we want a Countdown timer instead of an Elapsed timer, then we should just create a separate timer that counts down, with explicit semantics in its its API

That would be nice to have.  Using nanotime directly is a maintenance burden because any change can easily introduce comparison or unit conversion bugs.  It would be a nice goal if nanotime were only used used via unit tested classes that do not allow those two problems to occur.  The addition of the offset to Timer achieves this goal but a new class would be a nicer way to achieve the same goal.